### PR TITLE
cleanup dockerfile a bit

### PR DESCRIPTION
**Pull request by grosser**
mostly cherry-picked from https://github.com/zendesk/samson/pull/3187

@zendesk/bre @lox

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -34,7 +47,14 @@ COPY lib /app/lib
 
 # Assets
 COPY vendor/assets /app/vendor/assets
-RUN echo "takes 5 minute" && ./bin/decode_dot_env .env && RAILS_ENV=production PRECOMPILE=1 bundle exec rake assets:precompile
+RUN echo "takes 5 minute" && ./bin/decode_dot_env .env && RAILS_ENV=production PRECOMPILE=1 bundle exec rake assets:precompile 2>/dev/null
+
+# run as non-root, but allow it to access docker, use id so kubernetes non-root check passes
+RUN groupadd -r samson \
+  && useradd -r -g samson -d /app -u 1000 samson \
+  && chown -R samson:samson /app \
```
Comment (lox) [inline Dockerfile]: This will duplicate the entire source code layer unfortunately, check out the output of `docker history samson`.

**Comment by lox** [comment]
This image ends up with a HEAP of layers, I'd suggest combining as many of those `COPY` directives as possible, unlike `ADD` they can have multiple params.

**Comment by lox** [comment]
Out of interest, why do you silence the output of bundler and npm?

**Comment by grosser** [comment]
output was annoying, I wanted to get rid of it :D

**Comment by lox** [comment]
I view it as a good indicator that you've got a cache miss :) Also when things break it's helpful.

**Comment by grosser** [comment]
I'll check the layers, the image size was the same so I though this is fine
...

**Comment by lox** [comment]
This is what I ended up with after flattening the multi-stage down in to one: 

https://github.com/jwswj/samson/blob/master/.buildkite/Dockerfile

**Comment by lox** [comment]
For the Dockerfile in this PR I get:

```bash
docker build --tag samson --no-cache .  0.92s user 0.68s system 0% cpu 4:29.87 total

$ docker images samson
REPOSITORY          TAG                 IMAGE ID            SIZE
samson              latest              6e690d8b0d2d        914MB

$ docker history samson | wc -l
      40
```

For the one I linked to I get:

```bash
docker build -f .buildkite/Dockerfile --tag samson-bk --no-cache .  1.00s user 0.59s system 0% cpu 4:31.84 total

$ docker images samson-bk
REPOSITORY          TAG                 IMAGE ID            SIZE
samson-bk           latest              6cae0d5bc807        914MB

$ docker history samson-bk | wc -l
      29
```

Doesn't end up a huge difference in size or build time, but the latency of less layers in push and pull ends up being meaningful.

**Comment by grosser** [comment]
... removed the non-root for now and this is just a cleanup PR / reduces layers

File: Dockerfile [hunk 2]
```diff
@@ -1,30 +1,38 @@
...
+  && rm -rf /var/lib/apt/lists/* \
+  && curl -fsSL https://get.docker.com | bash -
 
 WORKDIR /app
 
 # Mostly static
-COPY config.ru /app/
-COPY Rakefile /app/
+COPY .ruby-version config.ru Rakefile .env.virtualbox /app/
 COPY bin /app/bin
```
Comment (sathishavm) [inline Dockerfile]: We no need to prepend '/app' everywhere, as we are in `WORKDIR /app`, just specify target file or folder.
eg: COPY bin bin
COPY .env.bootstrap .env

**Comment by grosser** [comment]
good catch :D

File: Dockerfile (hunk 2 above)
Comment (sathishavm) [reply Dockerfile]: Still some https://github.com/zendesk/samson/blob/5e97920ef5ed1b1529df91a91154da062730cd8f/Dockerfile#L39

**Comment by grosser** [comment]
got em all now :)

**Pull request merged**